File: solutions/typescript/resistor-color-trio/1/resistor-color-trio.ts
enum Resistor {
  black,
  brown,
  red,
  orange,
  yellow,
  green,
  blue,
  violet,
  grey,
  white,
}

type Color = keyof typeof Resistor

enum MetricPrefixes {
  ohms,
  kiloohms,
  megaohms,
  gigaohms,
}

export function decodedResistorValue([a, b, c]: Color[]): string {
  const rawValue = (Resistor[a] * 10 + Resistor[b]) * 10 ** Resistor[c]
  const [number, unit] = getNumberWithMetricPrefix(rawValue)

  return `${number} ${unit}`
}

function getNumberWithMetricPrefix(number: number): [number, string] {
  let unitLevel = 0

  while (number >= 1000) {
    number /= 1000
    unitLevel++
  }

  return [number, MetricPrefixes[unitLevel]]
}

File: solutions/elixir/resistor-color-trio/1/lib/resistor_color_trio.ex
defmodule ResistorColorTrio do
  @code_colors [
    black: 0,
    brown: 1,
    red: 2,
    orange: 3,
    yellow: 4,
    green: 5,
    blue: 6,
    violet: 7,
    grey: 8,
    white: 9
  ]

  @ohm_units [:ohms, :kiloohms, :megaohms, :gigaohms]

  @type ohm :: :ohms | :kiloohms | :megaohms | :gigaohms

  @doc """
  Calculate the resistance value in ohms from resistor colors
  """
  @spec label(colors :: [atom]) :: {number, ohm()}
  def label(colors) do
    [a, b, c] = Enum.take(colors, 3)
    with_metric_unit((@code_colors[a] * 10 + @code_colors[b]) * 10 ** @code_colors[c])
  end

  defp with_metric_unit(number, unit_level \\ 0)

  defp with_metric_unit(number, unit_level) when number >= 1000 do
    with_metric_unit(number / 1000, unit_level + 1)
  end

  defp with_metric_unit(number, unit_level) do
    {number, Enum.at(@ohm_units, unit_level)}
  end
end
